From d47b92fe0f94cfc56fab82ae06782766e5c765df Mon Sep 17 00:00:00 2001
From: Derek Schuff <dschuff@chromium.org>
Date: Wed, 11 Feb 2026 01:13:57 +0000
Subject: [PATCH 14/30] lemon: Fix declarations, and remove use of fork()

 - Fix Emscripten execution by avoiding fork(). Instead, consolidate the global
   state and reset it explicitly.
 - remove K&R declarations, they crash the compiler on wasm64
---
 MultiSource/Applications/lemon/lemon.c | 78 ++++++++++++++++++--------
 1 file changed, 54 insertions(+), 24 deletions(-)

diff --git a/MultiSource/Applications/lemon/lemon.c b/MultiSource/Applications/lemon/lemon.c
index 153f5b589..fde21b82b 100644
--- a/MultiSource/Applications/lemon/lemon.c
+++ b/MultiSource/Applications/lemon/lemon.c
@@ -331,30 +331,30 @@ struct state **State_arrayof(/*  */);
 void Configtable_init(/* void */);
 int Configtable_insert(/* struct config * */);
 struct config *Configtable_find(/* struct config * */);
-void Configtable_clear(/* int(*)(struct config *) */);
+void Configtable_clear(int(*)(struct config *));
 /****************** From the file "action.c" *******************************/
 /*
 ** Routines processing parser actions in the LEMON parser generator.
 */
 
+static struct action *action_freelist = 0;
 /* Allocate a new parser action */
 static struct action *Action_new(void){
-  static struct action *freelist = 0;
   struct action *new;
 
-  if( freelist==0 ){
+  if( action_freelist==0 ){
     int i;
     int amt = 100;
-    freelist = (struct action *)calloc(amt, sizeof(struct action));
-    if( freelist==0 ){
+    action_freelist = (struct action *)calloc(amt, sizeof(struct action));
+    if( action_freelist==0 ){
       fprintf(stderr,"Unable to allocate memory for a new parser action.");
       exit(1);
     }
-    for(i=0; i<amt-1; i++) freelist[i].next = &freelist[i+1];
-    freelist[amt-1].next = 0;
+    for(i=0; i<amt-1; i++) action_freelist[i].next = &action_freelist[i+1];
+    action_freelist[amt-1].next = 0;
   }
-  new = freelist;
-  freelist = freelist->next;
+  new = action_freelist;
+  action_freelist = action_freelist->next;
   return new;
 }
 
@@ -1398,14 +1398,14 @@ static void handle_D_option(char *z){
 /* The main program.  Parse the command line and do it... */
 int lemon_main(int argc,char **argv)
 {
-  static int version = 0;
-  static int rpflag = 0;
-  static int basisflag = 0;
-  static int compress = 0;
-  static int quiet = 0;
-  static int statistics = 0;
-  static int mhflag = 0;
-  static struct s_options options[] = {
+  int version = 0;
+  int rpflag = 0;
+  int basisflag = 0;
+  int compress = 0;
+  int quiet = 0;
+  int statistics = 0;
+  int mhflag = 0;
+  struct s_options options[] = {
     {OPT_FLAG, "b", (char*)&basisflag, "Print only the basis in report."},
     {OPT_FLAG, "c", (char*)&compress, "Don't compress the action table."},
     {OPT_FSTR, "D", (char*)handle_D_option, "Define an %ifdef macro."},
@@ -1423,11 +1423,11 @@ int lemon_main(int argc,char **argv)
   OptInit(argv,options,stderr);
   if( version ){
      printf("Lemon version 1.0\n");
-     exit(0); 
+     return 0;
   }
   if( OptNArgs()!=1 ){
     fprintf(stderr,"Exactly one filename argument is required.\n");
-    exit(1);
+    return 1;
   }
   memset(&lem, 0, sizeof(lem));
   lem.errorcnt = 0;
@@ -1445,10 +1445,10 @@ int lemon_main(int argc,char **argv)
 
   /* Parse the input file */
   Parse(&lem);
-  if( lem.errorcnt ) exit(lem.errorcnt);
+  if( lem.errorcnt ) return (lem.errorcnt);
   if( lem.nrule==0 ){
     fprintf(stderr,"Empty grammar.\n");
-    exit(1);
+    return 1;
   }
 
   /* Count and index the symbols of the grammar */
@@ -1515,7 +1515,6 @@ int lemon_main(int argc,char **argv)
     printf("                   %d states, %d parser table entries, %d conflicts\n",
       lem.nstate, lem.tablesize, lem.nconflict);
   }
-  exit(lem.errorcnt + lem.nconflict);
   return (lem.errorcnt + lem.nconflict);
 }
 /******************** From the file "msort.c" *******************************/
@@ -4836,8 +4835,7 @@ struct config *key;
 
 /* Remove all data from the table.  Pass each data to the function "f"
 ** as it is removed.  ("f" may be null to avoid this step.) */
-void Configtable_clear(f)
-int(*f)(/* struct config * */);
+void Configtable_clear(int(*f)(struct config *))
 {
   int i;
   if( x4a==0 || x4a->count==0 ) return;
@@ -4852,6 +4850,27 @@ int(*f)(/* struct config * */);
 #include <unistd.h>
 #include <sys/wait.h>
 #include <errno.h>
+static void lemon_reset() {
+  action_freelist = 0;
+  freelist = 0;
+  current = 0;
+  currentend = &current;
+  basis = 0;
+  basisend = &basis;
+  nDefine = 0;
+  azDefine = 0;
+  argv = 0;
+  op = 0;
+  errstream = 0;
+  plink_freelist = 0;
+  size = 0;
+  x1a = 0;
+  x2a = 0;
+  x3a = 0;
+  x4a = 0;
+  append_str(NULL, 0, 0, 0);
+}
+
 /* llvm-test supports only running program once,
  * we need to run it multiple times, because it only accepts
  * one input a time, and has a global state */
@@ -4863,6 +4882,16 @@ int main(int argc, char **argv)
 		 * meaningful timings */
 		for(i=1;i<argc;i++) {
 			int status;
+#ifdef __EMSCRIPTEN__
+			char *argv_child[] = {"lemon-child","-s",argv[i],NULL};
+			/* child */
+			fprintf(stdout,"Processing %s\n",mybasename(argv[i]));
+			lemon_reset();
+			status = lemon_main(3, argv_child);
+			if(status) {
+				fprintf(stderr,"Error while running on: %s\n",mybasename(argv[i]));
+			}
+#else
 			pid_t p = fork();
 			if(p == 0) {
 				char *argv_child[] = {"lemon-child","-s",argv[i],NULL};
@@ -4875,6 +4904,7 @@ int main(int argc, char **argv)
                                 fflush(stdout);
 				fprintf(stderr,"Error while running on: %s\n",mybasename(argv[i]));
 			}
+#endif
 		}
 	}
 	return 0;
-- 
2.53.0.1213.gd9a14994de-goog

